peer-signer/src/verifier.rs

```diff
@@ -0,0 +1,129 @@
+#[clap(version = "1.0")]
+pub struct Config {
+    // Node to ask for network state
+    #[clap(long, default_value = "http://127.0.0.1:9933")]
+    pub node: String,
+
+    /// Peer Id of the peer we want to verify.
+    #[clap(long)]
+    pub peer_id: String,
+
+    /// Message that the signature of we want to check.
+    #[clap(long)]
+    pub message: String,
+
+    /// Hex encoded public key of the peer.
+    #[clap(long)]
+    pub public_key: String,
+
+    /// Signature we want to check.
+    #[clap(long)]
+    pub signature: String,
+
+    /// Max block difference with head.
+    #[clap(long, default_value = "10")]
+    pub block_difference: u64,
+}
+
+async fn make_request_and_parse_result<T: serde::de::DeserializeOwned>(
+    client: &Client,
+    endpoint: &str,
+    body: Value,
+) -> T {
+    let mut response: Value = client
+        .post(endpoint)
+        .json(&body)
+        .send()
+        .await
+        .expect("Storage request has failed")
+        .json()
+        .await
+        .expect("Could not deserialize response as JSON");
+    let result = response["result"].take();
+    serde_json::from_value(result).expect("Incompatible type of the result")
+}
+
+fn construct_json_body(method_name: &str, params: Value) -> Value {
+    serde_json::json!({
+        "jsonrpc": "2.0",
+        "id": 1,
+        "method": method_name,
+        "params": params
+    })
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+struct Peer {
+    best_hash: String,
+    best_number: u64,
+    peer_id: String,
+    roles: String,
+}
+
+#[tokio::main]
+async fn main() {
+    let Config {
+        node,
+        peer_id,
+        message,
+        public_key,
+        signature,
+        block_difference,
+    } = Config::parse();
+
+    let req = construct_json_body("system_peers", Value::Null);
+    let block_req = construct_json_body("chain_getBlock", Value::Null);
+
+    let client = Client::new();
+    let res: Vec<Peer> = make_request_and_parse_result(&client, &node, req).await;
+    let block_res: Value = make_request_and_parse_result(&client, &node, block_req).await;
+
+    // For some reason in this rpc call result is returned in hex encoded decimal, so such ugly conversion is needed
+    let block_number = block_res["block"]["header"]["number"].as_str().unwrap();
+    let best_number = u64::from_str_radix(block_number.trim_start_matches("0x"), 16).unwrap();
+
+    if let Some(peer) = res.iter().find(|r| r.peer_id == peer_id) {
+        if peer.best_number + block_difference < best_number {
+            panic!(
+                "Peer is not up to date. Should have {:?} block number. Has {:?} block number",
+                best_number, peer.best_number
+            )
+        }
+        if best_number + block_difference < peer.best_number {
+            panic!("Peer is too far in the future. Should have {:?} block number. Has {:?} block number", best_number, peer.best_number)
+        }
```

Comment:
I see what is the value in that but for now we haven't thought about how to detect errors of this tool when using in pipeline. We probably don't want to run it by hand for every peer XD So for now I would stick with asserts and panics, as they are simplest to implement and are already done. In the future when we incorporate this tool so that a lot of peers can be verified automatically we can think about it again.